### ISSUE TITLE
Same flag for different commands

### ISSUE THREAD

**Issue by rmsj** [spf13/cobra#553]
How to add same flag for different command?
I've tried LocalNonPersistentFlags, LocalFlags, PersistentFlags.
Flags with the same name or shorthand do not work.
Viper fails to get the value.

**Comment by dnephin** [comment]
https://github.com/spf13/cobra#local-flag-on-parent-commands may be what you're looking for, but I'm not sure how it works with viper

**Comment by rmsj** [comment]
Hum I don't get it: why do I need to traverse children? I would have though my command have no children.
Command "abc" has flags x and y and command "def" has flags with same name. What is the parent child relationship here?

**Comment by bogem** [comment]
Can you please provide an example code of your problem?

**Comment by rmsj** [comment]
`// authorizeCmd represents the authorize command
var authorizeCmd = &cobra.Command{
	Use:   Authorize,
	Short: "Sends an authorization request to the central system.",
	Long: `Before the owner of an electric vehicle can start or stop charging, the Charge Point has to authorize the
operation. The Charge Point SHALL only supply energy after authorization.`,
	RunE:             AuthorizeRun,
	TraverseChildren: true,
}

func init() {
	RootCmd.AddCommand(authorizeCmd)

	authorizeCmd.Flags().StringP("auth", "a", "", "The RFID tag ID")
	viper.BindPFlag("auth", authorizeCmd.Flags().Lookup("auth"))
}`

This is one command with the flag 'auth'.
It works fine, until I add another command with same flag. Then it is just simply not able to read the value.

**Comment by lirao** [comment]
I'm wondering about the same thing too. The only way I can fix this is to add the flag to the parent command as a persistent flag.. but then i can't do different help messages for different subcommands.

**Comment by dturanski** [comment]
Same issue trying to compose commands.  Suppose you have commands b and c, and command  a executes b and c.  a needs to expose the a flags and b flags.  This is not easy to do without cut and paste.

**Comment by seanjoe** [comment]
I have the same issue. 
	rootCmd.PersistentFlags().BoolVar(&config.ShowDebugInfo, "debug", true, "Show debug info")
	rootCmd.AddCommand(prepareCmd)
	prepareCmd.PersistentFlags().BoolVar(&Auto, "auto", true, "help")
	rootCmd.AddCommand(buildCmd)
	buildCmd.PersistentFlags().BoolVar(&Auto, "auto", false, "help")
	rootCmd.AddCommand(deployCmd)
	deployCmd.PersistentFlags().BoolVar(&Auto, "auto", false, "help")
the default value auto was change by other subcommand
can use other method to avoid it?
I try use LocalFlags()  InheritedFlags() don't work

**Comment by alexisvisco** [comment]
Same issue. Annoying problem.

**Comment by pplcc** [comment]
I ran into the same problem.

**Comment by umarcor** [comment]
@lirao, @dturanski, @mcolburn, @pplcc can you please provide a MWE that reproduces your issue? None of the descriptions in this issue is detailed enough to reproduce.

**Comment by moqadir** [comment]
Use `DisableFlagParsing` and parse the flags in the sub command.

```
package main

import (
    "fmt"
    "os"

    "github.com/spf13/cobra"
)

func main() {

    rootCmd := &cobra.Command{
        Use:   "cobra",
        Short: "Cobra commander",
    }

    fooCmd := &cobra.Command{
        Use:                "foo",
        Short:              "Print foo message",
        Run:                runFoo,
        DisableFlagParsing: true,
    }

    barCmd := &cobra.Command{
        Use:                "bar",
        Short:              "Print bar message",
        Run:                runBar,
        DisableFlagParsing: true,
    }

    rootCmd.AddCommand(fooCmd)
    rootCmd.AddCommand(barCmd)
    if err := rootCmd.Execute(); err != nil {
        fmt.Println(err)
        os.Exit(1)
    }

}

func runFoo(cmd *cobra.Command, args []string) {
    cmd.Flags().String("print-string", "Default foo string", "Default foo string to print.")
    cmd.Flags().Parse(args)

    str, _ := cmd.Flags().GetString("print-string")
    fmt.Printf("Foo: %s\n", str)

}

func runBar(cmd *cobra.Command, args []string) {
    cmd.Flags().String("print-string", "Default bar string", "Default bar string to print.")
    cmd.Flags().Parse(args)

    str, _ := cmd.Flags().GetString("print-string")
    fmt.Printf("Bar: %s\n", str)
}
```

**Comment by shabeebqadir** [comment]
Use `DisableFlagParsing` and parse flags in the sub command.

```
package main

import (
    "fmt"
    "os"

    "github.com/spf13/cobra"
)

func main() {

    rootCmd := &cobra.Command{
        Use:   "cobra",
        Short: "Cobra commander",
    }

    fooCmd := &cobra.Command{
        Use:                "foo",
        Short:              "Print foo message",
        Run:                runFoo,
        DisableFlagParsing: true,
    }

    barCmd := &cobra.Command{
        Use:                "bar",
        Short:              "Print bar message",
        Run:                runBar,
        DisableFlagParsing: true,
    }

    rootCmd.AddCommand(fooCmd)
    rootCmd.AddCommand(barCmd)
    if err := rootCmd.Execute(); err != nil {
        fmt.Println(err)
        os.Exit(1)
    }

}

func runFoo(cmd *cobra.Command, args []string) {
    cmd.Flags().String("print-string", "Default foo string", "Default foo string to print.")
    cmd.Flags().Parse(args)

    str, _ := cmd.Flags().GetString("print-string")
    fmt.Printf("Foo: %s\n", str)

}

func runBar(cmd *cobra.Command, args []string) {
    cmd.Flags().String("print-string", "Default bar string", "Default bar string to print.")
    cmd.Flags().Parse(args)

    str, _ := cmd.Flags().GetString("print-string")
    fmt.Printf("Bar: %s\n", str)
}
```

**Comment by johnSchnake** [comment]
The given example works just fine for me; maybe something has changed in the past 2y though.

In addition, it isn't recommended that you define flags and parse within the Run command. By doing so you prevent that flag from being part of the help output/docs.

```
➜  ./tmp bar --print-string yay
Bar: yay
➜  ./tmp foo --print-string yay
Foo: yay
➜  ./tmp bar
Bar: Default bar string
➜  ./tmp foo
Foo: Default foo string
```

Closing due to the time since last comment and the example given doesn't demonstrate any bug. Feel free to reopen if this example wasn't representative and its still an ongoing issue.

**Comment by jurgen-weber-deltatre** [comment]
So, using this same example.. how can I get help to work for the subcommand?

> ./tmp var -h

just executes the code and doens't display the help/messages.

**Comment by m4x1202** [comment]
Stumbled over this issue as well. There is an imho better workaround explained in the [viper](https://github.com/spf13/viper) repository: https://github.com/spf13/viper/issues/375#issuecomment-578552586

**Comment by cclose** [comment]
I also recently tripped over this issue, and didn't see my solution posted, so thought I would share.

It seems the general consensus is that binding the pflags in the init() is an anti-pattern. In my case, I discovered this very late in a project, and moving my flag declarations out of init and into run would be a big lift. Lucky for me, I had abstracted things into a helper function that accepts FlagSets, and iterates the flag set, calling viper.BindPFlag and setting up some aliases:

```
func BindAndAlias(fs *pflag.FlagSet) {
	fs.VisitAll(func(f *pflag.Flag) {
		key := f.Name
		_ = viper.BindPFlag(key, f)
		if strings.Contains(key, "-") {
			dot := strings.ReplaceAll(key, "-", ".")
			viper.RegisterAlias(dot, key)
		}
	})
}
```

The end result for me was that I moved this call out of my init and into a PersistentPreRun:
```
	newCmd.PersistentPreRun = func(cmd *cobra.Command, args []string) {
		cli.BindAndAlias(newCmd.PersistentFlags())
		cli.BindAndAlias(newCmd.Flags())
	}
```

This makes sure that only the command that is ACTUALLY running is binding it's PFlags to Viper.
You do need to make sure that parent commands bind their flags to Viper as well if you want those to work. 
Kind of makes me overall want to re-think things... the architecture is a bit gross, but this was a quick fix to a project that already had a huge fan out of sub commands with overlapping flag sets on the leaf-nodes.
